sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java

```diff
@@ -28,63 +28,78 @@
 /** Tests for {@link ByteKeyRangeTracker}. */
 @RunWith(JUnit4.class)
 public class ByteKeyRangeTrackerTest {
-  private static final ByteKey START_KEY = ByteKey.of(0x12);
+  private static final ByteKey INITIAL_START_KEY = ByteKey.of(0x12);
+  private static final ByteKey NEW_START_KEY = ByteKey.of(0x13);
   private static final ByteKey MIDDLE_KEY = ByteKey.of(0x23);
   private static final ByteKey BEFORE_END_KEY = ByteKey.of(0x33);
   private static final ByteKey END_KEY = ByteKey.of(0x34);
-  private static final double RANGE_SIZE = 0x34 - 0x12;
-  private static final ByteKeyRange RANGE = ByteKeyRange.of(START_KEY, END_KEY);
+  private static final double INITIAL_RANGE_SIZE = 0x34 - 0x12;
+  private static final ByteKeyRange INITIAL_RANGE = ByteKeyRange.of(INITIAL_START_KEY, END_KEY);
+  private static final ByteKeyRange NEW_RANGE = ByteKeyRange.of(NEW_START_KEY, END_KEY);
 
   /** Tests for {@link ByteKeyRangeTracker#toString}. */
   @Test
   public void testToString() {
-    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(RANGE);
-    String expected = String.format("ByteKeyRangeTracker{range=%s, position=null}", RANGE);
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
+    String expected = String.format("ByteKeyRangeTracker{range=%s, position=null}", INITIAL_RANGE);
     assertEquals(expected, tracker.toString());
 
+    tracker.tryReturnRecordAt(true, INITIAL_START_KEY);
     tracker.tryReturnRecordAt(true, MIDDLE_KEY);
-    expected = String.format("ByteKeyRangeTracker{range=%s, position=%s}", RANGE, MIDDLE_KEY);
+    expected =
+        String.format("ByteKeyRangeTracker{range=%s, position=%s}", INITIAL_RANGE, MIDDLE_KEY);
+    assertEquals(expected, tracker.toString());
+  }
+
+  /** Tests for updating the start key to the first record returned. */
+  @Test
+  public void testUpdateStartKey() {
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
+
+    tracker.tryReturnRecordAt(true, NEW_START_KEY);
+    String expected =
+        String.format("ByteKeyRangeTracker{range=%s, position=%s}", NEW_RANGE, NEW_START_KEY);
     assertEquals(expected, tracker.toString());
   }
 
   /** Tests for {@link ByteKeyRangeTracker#of}. */
   @Test
   public void testBuilding() {
-    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(RANGE);
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
 
-    assertEquals(START_KEY, tracker.getStartPosition());
+    assertEquals(INITIAL_START_KEY, tracker.getStartPosition());
     assertEquals(END_KEY, tracker.getStopPosition());
   }
 
   /** Tests for {@link ByteKeyRangeTracker#getFractionConsumed()}. */
   @Test
   public void testGetFractionConsumed() {
```

Comment:
I expected a test of `getFractionConsumed` with a key that is NOT the initial start key, to ensure that progress gets and stays updated.